pkg/graveler/settings/manager.go

```diff
@@ -131,34 +186,73 @@ func (m *Manager) getFromStore(ctx context.Context, repositoryID graveler.Reposi
 		return nil, err
 	}
-	defer func() {
-		_ = reader.Close()
-	}()
+	defer reader.Close()
 	return io.ReadAll(reader)
 }
 
-// UpdateWithLock atomically gets a setting, performs the update function, and persists the setting to the store.
+// Update atomically gets a setting, performs the update function, and persists the setting to the store.
 // The settingTemplate parameter is used to determine the type passed to the update function.
-func (m *Manager) UpdateWithLock(ctx context.Context, repositoryID graveler.RepositoryID, key string, settingTemplate proto.Message, update func(proto.Message) error) error {
+func (m *KVManager) Update(ctx context.Context, repositoryID graveler.RepositoryID, key string, settingTemplate proto.Message, update updateFunc) error {
 	repo, err := m.refManager.GetRepository(ctx, repositoryID)
 	if err != nil {
 		return err
 	}
-	_, err = m.branchLock.MetadataUpdater(ctx, repositoryID, repo.DefaultBranchID, func() (interface{}, error) {
-		setting, err := m.GetLatest(ctx, repositoryID, key, settingTemplate)
-		if errors.Is(err, graveler.ErrNotFound) {
-			setting = proto.Clone(settingTemplate)
-		} else if err != nil {
-			return nil, err
-		}
-		logSetting(logging.FromContext(ctx), repositoryID, key, setting, "got repository-level setting")
-		err = update(setting)
-		if err != nil {
-			return nil, err
+
+	data := Settings{}
+	pred, err := m.getWithPredicate(ctx, &graveler.RepositoryRecord{RepositoryID: repositoryID, Repository: repo}, key, &data)
+	if err != nil && !errors.Is(err, graveler.ErrNotFound) {
+		return err
+	}
+	setting, err := m.GetLatest(ctx, repositoryID, key, settingTemplate)
+	if errors.Is(err, graveler.ErrNotFound) {
+		setting = proto.Clone(settingTemplate)
+	} else if err != nil {
+		return err
+	}
+
+	logSetting(logging.FromContext(ctx), repositoryID, key, setting, "got repository-level setting")
+	err = update(setting)
+	if err != nil {
+		return err
+	}
+	messageBytes, err := proto.Marshal(setting)
+	if err != nil {
+		return err
+	}
+	oldIdentifier := data.Identifier // Save old identifier for deletion
+	data.Identifier = fmt.Sprintf(settingsUniqueRelativeKey, m.committedBlockStoragePrefix, key, xid.NewWithTime(time.Now()))
+	obj := block.ObjectPointer{
+		StorageNamespace: string(repo.StorageNamespace),
+		Identifier:       data.Identifier,
+		IdentifierType:   block.IdentifierTypeRelative,
+	}
+	// Save to  blockstore
+	err = m.blockAdapter.Put(ctx, obj, int64(len(messageBytes)), bytes.NewReader(messageBytes), block.PutOpts{})
+	if err != nil {
+		return err
+	}
+
+	// Write key in KV Store
+	err = m.store.SetMsgIf(ctx, graveler.RepoPartition(&graveler.RepositoryRecord{RepositoryID: repositoryID, Repository: repo}), []byte(settingsPath(key)), &data, pred)
+	if err != nil {
+		// Delete file from blockstore
+		_ = m.blockAdapter.Remove(ctx, obj)
+		if errors.Is(err, kv.ErrPredicateFailed) {
+			err = fmt.Errorf("save setting in progress: %w", graveler.ErrPreconditionFailed)
 		}
```

Comment:
I think this error should be repeated. This error is due to an implementation detail, retrying here should mask this away from the user

Reply:
Do you suggest update should be retried?